### PR TITLE
feat: port clockskew support

### PR DESCRIPTION
**Pull request by ksegun**
The approach is similar to that used by the golang mongodb driver to maintain backwards compatibility

This should mitigate https://github.com/golang-jwt/jwt/issues/16

### CODE REVIEW

**Comment by oxisto** [comment]
> The approach is similar to that used by the golang mongodb driver to maintain backwards compatibility
> 
> This should mitigate #16

Thanks for the contribution, very much welcome, this is a feature that was on the wish list for a long time. This is definitely something I also wanted to explore as part of the new validator helper approach. I promise that I will find some time on the upcoming weekend to have a look at this, unfortunately we are quite in a end-of-the-year rush here. I also want to align this with the work I have previously started as well.

**Comment by mfridman** [comment]
Agree with much of what @oxisto said. (I'm also a bit swamped)

My only 2 cents is we should take our time and think through whether this is the correct API and ensure this is a backwards compatible change. (not saying it's not, just that we think through this).

We should also use functional options instead of a variadic pointer to a struct.

**Comment by ksegun** [comment]
Agree, we would prefer to have functional options for those of us stuck on the v4 preview in the old library we are quite happy with them. 

As we evaluated our options on whether to port to a different library with skew support or this, this was a trade-off that we thought could be adequate until functional options are made available and there could be far easier ways as well now that I consider it e.g. an environment variable.

We fully support whatever approach is decided on.

File: claims.go
Comment (oxisto) [inline claims.go:12]: As @mfridman already mentioned, would it be possible to turn these into functional style options? Similar to what we have in https://github.com/golang-jwt/jwt/blob/main/parser_option.go

File: claims_option.go
Comment (oxisto) [inline claims_option.go:6]: I would call this `ValidatorOptions`, similar to the `ParserOptions` we have

**Review by oxisto — CHANGES_REQUESTED** [review]

File: claims_option.go
Comment (ksegun) [reply claims_option.go]: Renamed

File: claims.go
Comment (ksegun) [reply claims.go]: The `WithLeeway` function is functional style https://github.com/golang-jwt/jwt/pull/139/files#diff-72f644d282bce7e136d0f762de1b9d5f3e447bfa88e5e0c4974ac2479107ddc9R33-R37 

It is far easier to set this as an attribute of the parser which is what the old library did, but it is not a backward compatible change and given how the `Valid` function is called (https://github.com/golang-jwt/jwt/pull/139/files#diff-83eb8e32639d01cf443d6d8bde24c1c8be78766090d8c5f8586c36250cfedca6R87)  I am frankly unsure how you make the claims validation functional style, however, I would be happy to make the change with your input.

**Comment by ksegun** [comment]
@oxisto this is now all functional style

**Comment by oxisto** [comment]
> @oxisto this is now all functional style

Nice! I will have a detailed look on this in the upcoming days. Thanks for the improvements

File: validator_option.go
Comment (oxisto) [inline validator_option.go:11]: Suggestion: This type should be called `Validator` or rather `validator`. Additionally, I would make this type unexported, at least for now, until our API is more stabilised.

File: validator_option.go
Comment (oxisto) [inline validator_option.go:8]: maybe ValidationOption instead? Probably a toin coss at some point. But the way I see it, this function MUST be exported (or must it) so we should choose a name we are happy with.

cc @mfridman

File: parser.go
Comment (oxisto) [inline parser.go:26]: validationOptions?

File: claims.go
Comment (oxisto) [inline claims.go:273]: `skew` would probably be more in line with the other short argument names

File: claims.go
Comment (oxisto) [inline claims.go:263]: Wouldn't it be more logically to Add/Sub the clock skew from `now`? Same goes for verifyNbf

File: claims.go
Comment (oxisto) [inline claims.go:12]: for the sake of brevity and consistency it should be `opts`

**Review by oxisto — COMMENTED** [review]
Just some minor things, but I want to have at least a small discussion on naming before we merge this (as this impacts the public API).

Good job on making this non-breaking!

cc @mfridman

File: claims.go
Comment (ksegun) [reply claims.go:263]: The math is sort of backward, decided to leave it alone. Unit tests confirm this is accurate.

File: validator_option.go
Comment (ksegun) [reply validator_option.go]: Only needs to be exported if we want to allow usage in the call to Valid outside of the parser.  Changed so nothing is exported except the `WithLeeway` function in the parser.

Comment (oxisto) [reply validator_option.go]: Good idea. Probably the best approach until this is stabilized from an API point of view.

File: claims.go
Comment (oxisto) [reply claims.go]: Should we call it leeway as a parameter then? Similar to the validator option. Then it would also make sense to add it to exp? What do you think?

Comment (ksegun) [reply claims.go]: Yes, we could I used skew here because that is typically how it is described in the literature and I felt it would make it clear in the code and retained WithLeeway because that is how the original library had it.

**Review by oxisto — DISMISSED** [review]
LGMT. This is / should be 100 % backwards compatible and fixes a long-wanted feature. Furthermore, it allows us to go into the direction of a validation helper that fine-tunes validation of the parser, also completely backwards compatible and incrementally. All new types are unexported, so we are free to change / stabilize the API. All function signature changes are variadic.

I definitely want to get some more votes on this before we merge though

cc @golang-jwt/maintainers

File: validator_option.go
Comment (mfridman) [inline validator_option.go:7]: I think this is a typo, s/`*ValidatorOptions`/`*validator`

File: claims.go
Comment (mfridman) [inline claims.go:12]: I think we should add a comment, either on this interface (or maybe even better) around the `validationOption` explaining why it is not an exported identifier. 

Looking at this method signature I'd be a little confused.

**Review by mfridman — DISMISSED** [review]
lgtm.

**Comment by oxisto** [comment]
@mfridman I have tweaked the documentation. I think we can merge it then? We are still free to change the internal names in the future.

**Review by oxisto — APPROVED** [review]

**Review by mfridman — APPROVED** [review]
Sounds good to me. 

Outside the scope of this PR, we might want to revisit all these functional options in a future breaking release. We should start a discussion about that?

**Comment by oxisto** [comment]
> Sounds good to me.
> 
> Outside the scope of this PR, we might want to revisit all these functional options in a future breaking release. We should start a discussion about that?

Ideally, we already started them in a way that we are happy with so not too many breaking changes are necessary. That is the main reason why I wanted to keep most of the validation structs and functions unexpected, so we can easily change them again internally.

File: claims.go
Comment (kolatat) [reply claims.go:15]: Since the interface changed to include an unexported type, it is impossible to implement user-defined `Claims` (and `Token` since it contains `Claims`)?

Comment (oxisto) [reply claims.go:15]: You should be able to implement user-defined `Claims` by embedding either `RegisteredClaims` or the old `StandardClaims`. It could be impossible to completely implement a different claims structure that does not re-use `RegisteredClaims`, but that is not intended anyway.

See our example here https://github.com/golang-jwt/jwt/blob/d489c99d3e8a82bd47dcb3e4880414f7f747eaa7/example_test.go#L33-L39

Comment (gustavosbarreto) [reply claims.go:15]: This breaks backwards compatibility for anyone using older `v3.x.y` and `dgrijalva/jwt-go`:

```
cannot use &"github.com/golang-jwt/jwt/v4".MapClaims{} (type *"github.com/golang-jwt/jwt/v4".MapClaims) as type "github.com/golang-jwt/jwt".Claims in field value:
	*"github.com/golang-jwt/jwt/v4".MapClaims does not implement "github.com/golang-jwt/jwt".Claims (wrong type for Valid method)
		have Valid(..."github.com/golang-jwt/jwt/v4".validationOption) error
		want Valid() error
```
Can you please consider adding a new function (`ValidWithOptions`) instead of changing `Valid()` funtion signature?

**Pull request merged**
